internal/server/singleprocess/service_instance.go

```diff
@@ -18,8 +21,43 @@ func (s *service) ListInstances(
 	var err error
 	switch scope := req.Scope.(type) {
 	case *pb.ListInstancesRequest_DeploymentId:
-		result, err = s.state.InstancesByDeployment(scope.DeploymentId, nil)
+		// For more complex deployments, sometimes it takes a bit longer for
+		// the instance to respond. This blocks on a response when querying
+		// a deployment by id, and lets the user know it is taking a while
+		//
+		// The default case is no timeout, no blocking. Make a request and return
+
+		timeout := "0"
+		if req.ConnectTimeout != "" {
+			timeout = req.ConnectTimeout
+		}
+		connectTimeout, err := time.ParseDuration(timeout)
+		if err != nil {
+			return nil, err
+		}
+
+		if connectTimeout > 0 {
+			_, cancel := context.WithTimeout(ctx, connectTimeout)
+			defer cancel()
+		}
+
+		for {
+			ws := memdb.NewWatchSet()
+			result, err = s.state.InstancesByDeployment(scope.DeploymentId, ws)
+
+			if err != nil {
+				return nil, err
+			}
+
+			if len(result) > 0 || connectTimeout == 0 {
```

Comment:
building on the above, can change the 2nd part here to `req.ConnectTimeout == ""`